Mage.Sets/src/mage/cards/b/BloodSpatterAnalysis.java

```diff
@@ -0,0 +1,108 @@
+import mage.target.common.TargetOpponentsCreaturePermanent;
+
+import java.util.UUID;
+
+/**
+ * @author Cguy7777
+ */
+public final class BloodSpatterAnalysis extends CardImpl {
+
+    public BloodSpatterAnalysis(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ENCHANTMENT}, "{B}{R}");
+
+        // When Blood Spatter Analysis enters the battlefield, it deals 3 damage to target creature an opponent controls.
+        Ability ability = new EntersBattlefieldTriggeredAbility(new DamageTargetEffect(3));
+        ability.addTarget(new TargetOpponentsCreaturePermanent());
+        this.addAbility(ability);
+
+        // Whenever one or more creatures die, mill a card and put a bloodstain counter on Blood Spatter Analysis.
+        // Then sacrifice it if it has five or more bloodstain counters on it.
+        // When you do, return target creature card from your graveyard to your hand.
+        this.addAbility(new BloodSpatterAnalysisTriggeredAbility());
+    }
+
+    private BloodSpatterAnalysis(final BloodSpatterAnalysis card) {
+        super(card);
+    }
+
+    @Override
+    public BloodSpatterAnalysis copy() {
+        return new BloodSpatterAnalysis(this);
+    }
+}
+
+class BloodSpatterAnalysisTriggeredAbility extends TriggeredAbilityImpl {
+
+    BloodSpatterAnalysisTriggeredAbility() {
+        super(Zone.BATTLEFIELD, new MillCardsControllerEffect(1));
+        this.addEffect(new AddCountersSourceEffect(CounterType.BLOODSTAIN.createInstance()).concatBy("and"));
+
+        ReflexiveTriggeredAbility returnAbility = new ReflexiveTriggeredAbility(new ReturnFromGraveyardToHandTargetEffect(), false);
+        returnAbility.addTarget(new TargetCardInYourGraveyard(StaticFilters.FILTER_CARD_CREATURE_YOUR_GRAVEYARD));
+
+        this.addEffect(new ConditionalOneShotEffect(
+                new DoWhenCostPaid(returnAbility, new SacrificeSourceCost(), null, false),
+                new SourceHasCounterCondition(CounterType.BLOODSTAIN, 5))
+                .setText("Then sacrifice it if it has five or more bloodstain counters on it. When you do, return target creature card from your graveyard to your hand"));
+
+        setTriggerPhrase("Whenever one or more creatures die, ");
+    }
+
+    private BloodSpatterAnalysisTriggeredAbility(final BloodSpatterAnalysisTriggeredAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE_BATCH;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        ZoneChangeBatchEvent zBatchEvent = (ZoneChangeBatchEvent) event;
+        if (zBatchEvent == null) {
+            return false;
+        }
+
+        for (ZoneChangeEvent zEvent : zBatchEvent.getEvents()) {
+            if (zEvent.getFromZone().match(Zone.BATTLEFIELD) && zEvent.getToZone().match(Zone.GRAVEYARD)) {
```

Comment:
Must use helper method zEvent. isDiesEvent. Also I recommends to use equals, objects.equals or == for zones compare. “Match” method used by game engine to search zones like Zone.ALL.

Reply:
Good points, thank you. The existence of `isDiesEvent()` must've slipped past me.